### ISSUE TITLE
Split NuGet package into few child packages to ease dependencies, lower minimum target framework

### ISSUE THREAD

**Issue by abatishchev** [jwt-dotnet/jwt#76]
The idea is to extract few child packages:

- JwtNet.Json.NewtonSoft for JsonNet based implementation of Json (de)serialization
- JwtNet.Json.ServiceStack for ServiceStack.Text based implementations
- JwtNet.Cryptography.RS256 for System.Security.Cryptography based implementation of RS256Algorithm

**Comment by nbarbettini** [comment]
I like this idea, because it will mean that the "base" functionality can exist in a package that supports the widest number of platforms/devices. The only thing to figure out is how to structure the package graph to be both flexible and have easy discoverability.

For example, if you do `install-package JWT`, should you get `JWT.Json.Newtonsoft` automatically? It would be less confusing if you did, but if it's declared as a dependency of `JWT`, you won't be able to remove it.

One solution I've used in the past is to push all of the interfaces (and maybe base functionality) into a package called `JWT.Abstractions` or `JWT.Core`, and then make the top-level `JWT` package just be an empty metapackage that serves as an easy target for discoverability. That would look like:

```
JWT (empty package, just dependencies)
  - depends on JWT.Abstractions
  - depends on JWT.Json.Newtonsoft
    - depends on JWT.Abtractions
```

So, `install-package JWT` gives you the default packages. If you want to customize the installation, you can skip `JWT` and install `JWT.Abtractions` directly plus whatever supporting packages you want.

Any thoughts?

**Comment by abatishchev** [comment]
Sounds like a plan! I would just avoid calling it Core to avoid confusion with .NET Core. At work I usually call such project Contracts but we can make up something different.

Should we bump major version to 3.x since it'll be one more big breaking change.

**Comment by nbarbettini** [comment]
> I would just avoid calling it Core to avoid confusion with .NET Core

Agreed 😄 

> Should we bump major version to 3.x since it'll be one more big breaking change.

I'm fine with that, would you want to fully drop the old static interface too?

**Comment by abatishchev** [comment]
Yes, I think that would be the right time to do it.

**Comment by zemien** [comment]
This looks like a great, easy-to-use library. I notice @abatishchev has started porting it to .Net Standard. Is there anything I can help with that process? I need something similar for my .Net Core project and I would rather help get this production-ready than do my own personal fork.

**Comment by abatishchev** [comment]
hi @zemien, the port is done (committed but not published yet). hey @nbarbettini, can you please pack and push to nuget?

**Comment by nbarbettini** [comment]
@zemien see the latest prerelease version (3.x series) on NuGet. That supports .NET Standard already. 👍

**Comment by zemien** [comment]
Awesome, I see it now, it's great work!

**Comment by nbarbettini** [comment]
Is this still something you'd like @abatishchev?

**Comment by abatishchev** [comment]
I'd park this since it seems there is no demand on implementations other than Json.Net. When/if such demand will be called out, we can come back on this. What do you think? Makes sense?

**Comment by nbarbettini** [comment]
> When/if such demand will be called out, we can come back on this. What do you think? Makes sense?

Yep, I agree. Doesn't seem to be a strong need to do this right away.

**Comment by carlittle** [comment]
I ended up forking this and creating my own build.  I had a need to use a Newtonsoft.Json that was no higher than 6.0.8 due to other dependencies on it.  The easiest way for me to do that was to not get the package from NuGet and buidl it myself.  I also built it with .NET 4.5.  

The build was painless.  Just changed the version of Newtonsoft that I was using and compiled without a hitch.

**Comment by abatishchev** [comment]
Do you know how to specify the minimal version of a dependency when you don't have nuspec but Core project file? I'd love to put Json.Net v6.0.3 as the dependency. Currently it puts referenced version by default and we didn't find a way to change that.

**Comment by abatishchev** [comment]
@carlittle any ideas on this?

**Comment by psylenced** [comment]
Just a small follow up on your targeting. I noticed you needed to bump up the version to 4.6.x due to `RSA.SignData`, however you targeted 4.6.2 which forced my 4.6.1 project into netstandard which I didn't want.

Is there any reason why you need 4.6.2 specifically? If not you should target the minimum required version (being `net46`).

**Comment by abatishchev** [comment]
@psylenced please open a new issue

**Comment by psylenced** [comment]
Issue created - #141

**Comment by abatishchev** [comment]
No plans to implement this due to no demand
